incubator/virtualcluster/pkg/syncer/controllers/pod/checker.go

```diff
@@ -0,0 +1,158 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pod
+
+import (
+	"context"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+
+	mc "github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/syncer/mccontroller"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/syncer/conversion"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/syncer/reconciler"
+	"sync"
+)
+
+// StartPeriodChecker starts the period checker for data consistency check. Checker is
+// blocking so should be called via a goroutine.
+func (c *controller) StartPeriodChecker(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	// Start a loop to periodically check if pods keep consistency between super
+	// master and tenant masters.
+	wait.Until(c.checkPods, c.periodCheckerPeriod, stopCh)
+}
+
+// checkPods checks to see if pods in super master informer cache and tenant master
+// keep consistency.
+func (c *controller) checkPods() {
+	clusterNames := c.multiClusterPodController.GetClusterNames()
+	if len(clusterNames) == 0 {
+		klog.Infof("tenant masters has no clusters, give up period checker")
+		return
+	}
+
+	wg := sync.WaitGroup{}
+
+	for _, clusterName := range clusterNames {
+		cluster := c.multiClusterPodController.GetCluster(clusterName)
+		if cluster == nil {
+			klog.Errorf("failed to locate cluster %s", clusterName)
+			continue
+		}
+
+		wg.Add(1)
+		go func(clusterName string, cluster mc.ClusterInterface) {
+			defer wg.Done()
+			c.checkPodsOfCluster(clusterName, cluster)
+		}(clusterName, cluster)
+	}
+	wg.Wait()
+
+	pPods, err := c.podLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("error listing pods from super master informer cache: %v", err)
+		return
+	}
+
+	for _, pPod := range pPods {
+		clusterName, vNamespace := conversion.GetVirtualOwner(pPod)
+		if len(clusterName) == 0 || len(vNamespace) == 0 {
+			continue
+		}
+
+		vPodObj, err := c.multiClusterPodController.Get(clusterName, vNamespace, pPod.Name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				if pPod.DeletionTimestamp != nil {
+					// pPod is under deletion, waiting for UWS bock populate the pod status.
+					continue
+				}
+				// vPod not found and pPod not under deletion, we need to delete pPod manually
+				deleteOptions := metav1.NewDeleteOptions(*pPod.DeletionGracePeriodSeconds)
```

Comment:
if pPod. DeletionTimestamp == nil. `pPod.DeletionGracePeriodSeconds` also a nil pointer. There is a panic.

using `pPod.Spec.TerminationGracePeriodSeconds` instead. if still nil pointer. just delete with 0 grace period seconds